### PR TITLE
docs: update db migration, environment and restful api practicesC

### PR DESCRIPTION
**Pull request by vdhieu**
#### What's this PR do?
Add best practices for
- [x] RestFul API
- [x] Database migration
- [x] Environment

### CODE REVIEW

**Review by zlatanpham — APPROVED** [review]
LGTM. Let @hieuphq take a quick look

File: engineering/backend/restful.md [hunk 1]
```diff
@@ -11,7 +11,7 @@ The following are the most important terms related to REST APIs.
 
 ## Convention
 
-- We use Swagger for REST API documentation.
+- We prefer Swagger for REST API documentation. Alternative can be postman
```
Comment (hieuphq) [inline engineering/backend/restful.md:14]:
```suggestion
- We prefer Swagger for REST API documentation.
```
we can use a tool to generate postman file from swagger

File: engineering/backend/restful.md [hunk 2]
```diff
@@ -33,27 +33,65 @@ Always version your API. Version via the URL, not via headers. Versioning APIs a
...
 GET /companies?page=23&size=50
 ```
 
-Sort:
+**Cursor pagination**
+
+- `cursor` - is the reference element in the list
+- `size` - number of items per page
+
+- The API should return `previousCursor` and `nextCursor`
```
Comment (hieuphq) [inline engineering/backend/restful.md:55]: should describe how we structure the cursor from the backend and use from client

File: engineering/backend/db-migration.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+- Every database changes will be documented incrementally, as small as possible.
+- The script should include unique sequence number or timestamp
+- Each migration must include up (migrate) and down (rollback).
+- Store schema migration history in database
+- All migration files are tracked in version control and saved in single place
+
+For example, we want to add indexes to user tables the migration file(s) would be: `20221010113000_add_indexes_to_users.up.sql` and `20221010113000_add_indexes_to_users.down.sql`
+
+
```
Comment (hieuphq) [inline engineering/backend/db-migration.md:19]: remove the empty line

File: engineering/backend/db-migration.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
+# Database migration practice
+
+> ## Backup Your Data
+> To avoid any potential data loss, before you start the migration, make sure you have a backup version of current database so you'll be able to restore if somethings don't go according to the plan.
+
+## Schema Migration
+The way you managing the database changes. It could be schema changes (e.g. adding/remove column, table) or data fixes (e.g. change data type of the column from json to jsonb)
+
+### The concept
+
```
Comment (hieuphq) [inline engineering/backend/db-migration.md:10]: remove the empty line

File: engineering/backend/db-migration.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
+# Database migration practice
+
```
Comment (hieuphq) [inline engineering/backend/db-migration.md:2]: remove the empty line

**Review by hieuphq — CHANGES_REQUESTED** [review]

File: engineering/backend/restful.md [hunk 6]
```diff
@@ -33,27 +33,65 @@ Always version your API. Version via the URL, not via headers. Versioning APIs a
...
+
+Example:
 ```
 GET /companies?page=23&size=50
 ```
 
-Sort:
+**Cursor pagination**
+
+- `cursor` - is the reference element in the list. Best brattice for choosing the cursor is only sort by unique sequential column, create new column which is concatenate columns if you have to order by multiple none unique columns.
```
Comment (hieuphq) [inline engineering/backend/restful.md:52]:
```suggestion
- `cursor` - is the reference element in the list. The best practice for choosing the cursor is only sorting by a unique sequential column, and creating a new concatenate column if you have to order by multiple none unique columns.
```

File: engineering/backend/restful.md [hunk 7]
```diff
@@ -33,27 +33,65 @@ Always version your API. Version via the URL, not via headers. Versioning APIs a
...
-GET /companies?time=fromTime-toTime
+# filter by category
+GET /companies?category=software
+
+# filter by locations
+GET /companies?location=saigon&location=hanoi
+
+# Filter by time range
+GET /companies?start=1662873449&end=1662873449
+GET /companies?start=2022-09-10T05:10:55.308Z&end=2022-09-11T05:10:55.308Z
```
Comment (hieuphq) [inline engineering/backend/restful.md:91]: how can we filter 2+ DateTime fields with this convention?

File: engineering/environment.md [hunk 8]
```diff
@@ -31,3 +31,17 @@ Staging is set up precisely like production. No changes to the production enviro
...
 
 This set helps developers to come up with the best technical solutions and resolve issues quickly, and users will only ever notice a new release by the product's increased awesomeness.
+
+
+## Environment variables
+- All environment must have exact list of variables
+- Your code should support load env variable from both `.env` file or from the `process`
+- In local development we use `.env` file to store all required environment variable. 
+- In production we use platform secret manager to set environment variables
+- `Do not` commit your `.env` file or hard coded secret variables
```
Comment (hieuphq) [inline engineering/environment.md:41]:
```suggestion
- **Do not** commit your `.env` file or hard-coded secret variables. We must set up our project in the `.gitignore` file.
```

File: engineering/environment.md [hunk 9]
```diff
@@ -31,3 +31,17 @@ Staging is set up precisely like production. No changes to the production enviro
...
+- Variant of `.env` file can be
+  - `.env.example` can commit t source code store default public variables, all private or secret values must not included in this file
```
Comment (hieuphq) [inline engineering/environment.md:43]:
```suggestion
  - `.env.example` can commit source code store default public variables, all private or secret values must not include in this file
```

File: engineering/backend/restful.md [hunk 10]
```diff
@@ -33,27 +33,65 @@ Always version your API. Version via the URL, not via headers. Versioning APIs a
...
+
 ```
-GET /companies?time=fromTime-toTime
+# filter by category
+GET /companies?category=software
+
+# filter by locations
+GET /companies?location=saigon&location=hanoi
+
+# Filter by time range
```
Comment (hieuphq) [inline engineering/backend/restful.md:89]: It makes me confused about the markdown convention. As we knew, One # character for Heading 1. We should update the way we index the headings.
```suggestion
#### Filter by category
GET /companies?category=software

#### Filter by locations
GET /companies?location=saigon&location=hanoi

#### Filter by time range
```

**Review by hieuphq — CHANGES_REQUESTED** [review]

**Review by monotykamary — CHANGES_REQUESTED** [review]
Just some nits; otherwise, LGTM 👍

For extras, specifically for environment, it might be worthwhile to mention local file ignoring with `$GIT_DIR/info/exclude` or `git update-index --skip-worktree <file-list>`.

**Pull request closed without merging**